src/app/clusters/operational-state-server/operational-state-server.h

```diff
@@ -33,54 +33,96 @@ class Uncopyable
 {
 public:
     /**
-     * Creates an operational state cluster instance. The Init() function needs to be called for this instance to be registered and
-     * called by the interaction model at the appropriate times.
+     * Creates an operational state cluster instance. The Init() function needs to be called for this instance
+     * to be registered and called by the interaction model at the appropriate times.
+     * @param aDelegate A pointer to the delegate to be used by this server.
+     * Note: the caller must ensure that the delegate lives throughout the instance's lifetime.
      * @param aEndpointId The endpoint on which this cluster exists. This must match the zap configuration.
-     * @param aClusterId The ID of the operational state aliased cluster to be instantiated.
+     * @param aClusterId The ID of the operational state derived cluster to be instantiated.
      */
-    OperationalStateServer(EndpointId aEndpointId, ClusterId aClusterId) :
-        CommandHandlerInterface(MakeOptional(aEndpointId), aClusterId),
-        AttributeAccessInterface(MakeOptional(aEndpointId), aClusterId)
-    {
-
-        mEndpointId = aEndpointId;
-        mClusterId  = aClusterId;
-    }
+    Instance(Delegate * aDelegate, EndpointId aEndpointId, ClusterId aClusterId);
 
-    ~OperationalStateServer() override {}
+    ~Instance() override;
 
     /**
-     * Init the operational state server.
-     * This function must be called after defining a OperationalStateServer class object.
-     * @param void
-     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     * Initialise the operational state server instance.
+     * This function must be called after defining an Instance class object.
+     * @return Returns an error if the given endpoint and cluster ID have not been enabled in zap or if the
+     * CommandHandler or AttributeHandler registration fails, else returns CHIP_NO_ERROR.
      */
     CHIP_ERROR Init();
 
+    // Attribute setters
+    /**
+     * Set operational phase.
+     * @param phase The operational phase that should now be the current one.
+     */
+    CHIP_ERROR SetCurrentPhase(const app::DataModel::Nullable<uint8_t> & phase);
+
     /**
-     * Shut down the operational state server.
-     * This function must be called before destroying a OperationalStateServer class object.
-     * @param void
+     * Set countdown time.
+     * @param time The countdown time that should now be the current one.
      */
-    void Shutdown();
+    CHIP_ERROR SetCountdownTime(const app::DataModel::Nullable<uint32_t> & time);
 
+    /**
+     * Set current operational state.
+     * @param opState The operational state that should now be the current one.
+     */
+    CHIP_ERROR SetOperationalState(uint8_t opState);
+
+    /**
+     * Set operational error.
+     * @param opErrState The new operational error.
+     */
+    CHIP_ERROR SetOperationalError(const GenericOperationalError & opErrState);
```

Comment:
I have removed this setter. See [this comment](https://github.com/project-chip/connectedhomeip/pull/28577#discussion_r1291408696).

However, I have kept the `OperationalError` as a member, managed by the Instance. I think that 128 bytes is a cheap price to pay for having the attribute entirely managed by the SDK. What do you think?

Reply:
128 bytes is not nothing, since we are within a few KB of hitting RAM limits on a bunch of chips.  We've certainly put in nontrivial work to save tens of bytes of RAM elsewhere.... But OK.

Reply:
This is the only outstanding conversation. I am going to resolve it and we can add an issue against this later if we can come up with a sensible way to save these 128 bytes.